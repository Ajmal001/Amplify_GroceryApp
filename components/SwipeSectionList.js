import React from "react";
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableHighlight,
  View,
  Button,
} from "react-native";
import { MaterialCommunityIcons } from "@expo/vector-icons";
import { SwipeListView } from "react-native-swipe-list-view";
import { mainColor, productCategory, secondaryColor } from "../utils/helpers";

function SwipeSectionList({
  listData,
  deleteProduct,
  navigateToEditProduct,
  toggleProduct,
  toBuyView,
}) {
  const closeRow = (rowMap, rowKey) => {
    if (rowMap[rowKey]) {
      rowMap[rowKey].closeRow();
    }
  };

  const deleteRow = (rowMap, rowKey) => {
    closeRow(rowMap, rowKey);
    deleteProduct(rowKey);
  };
  const goToEditProduct = (rowMap, product) => {
    closeRow(rowMap, product.key);
    navigateToEditProduct(product);
  };
  const onToggle = async (product) => {
    toggleProduct(product);
  };
  const onRowDidOpen = (rowKey) => {
    // console.log("This row opened", rowKey);
  };

  const renderItem = (data) => (
    <View>
      <TouchableHighlight
        onPress={() => onToggle(data.item)}
        style={data.item.toBuy? styles.rowFront : [styles.rowFront, styles.notSelected]}
        underlayColor={"#AAA"}
      >
        <View style={styles.rowAlign}>
          {toBuyView ? (
            <MaterialCommunityIcons
              style={styles.rowIcon}
              name={
                data.item.checked
                  ? "checkbox-marked-circle"
                  : "checkbox-blank-circle-outline"
              }
              size={20}
              color={mainColor}
            />
          ) : (
            <MaterialCommunityIcons
              style={styles.rowIcon}
              name={
                data.item.toBuy
                  ? "cart-outline"
                  : "cart-off"
              }
              size={20}
              color={'grey'}
            />
          )}

          <Text style={!data.item.toBuy ? {color: '#4d4c4c'}: {color: 'black'}}>
            {" "}
            {data.item.name}{" "}
            {data.item.quantity !== 0 && (
              <Text>
                {" "}
                ({data.item.quantity} {data.item.unit}){" "}
              </Text>
            )}
          </Text>
        </View>
      </TouchableHighlight>
    </View>
  );

  const renderHiddenItem = (data, rowMap) => (
    <View style={styles.rowBack}>
      <TouchableOpacity
        style={[styles.backRightBtn, styles.backRightBtnLeft]}
        onPress={() => goToEditProduct(rowMap, data.item)}
      >
        <MaterialCommunityIcons name="pencil" size={20} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[styles.backRightBtn, styles.backRightBtnRight]}
        onPress={() => deleteRow(rowMap, data.item.key)}
      >
        <MaterialCommunityIcons name="delete-outline" size={20} color="white" />
      </TouchableOpacity>
    </View>
  );

  const renderSectionHeader = ({ section }) => (
    <View style={styles.sectionHeader}>
      <MaterialCommunityIcons
        style={styles.sectionIcon}
        name={productCategory[section.title].picture}
        size={20}
      />
      <Text style={styles.sectionText}>{section.title}</Text>
    </View>
  );

  return (
    <SwipeListView
      useSectionList
      sections={listData}
      renderItem={renderItem}
      renderHiddenItem={renderHiddenItem}
      renderSectionHeader={renderSectionHeader}
      leftOpenValue={75}
      rightOpenValue={-150}
      previewRowKey={"0"}
      previewOpenValue={-40}
      previewOpenDelay={3000}
      onRowDidOpen={onRowDidOpen}
    />
  );
}

export default SwipeSectionList;

const styles = StyleSheet.create({
  container: {
    backgroundColor: "white",
    flex: 1,
  },
  sectionHeader: {
    flexDirection: "row",
    marginLeft: 10,
  },
  sectionText: {
    color: secondaryColor,
    fontWeight: "bold",
  },
  sectionIcon: {
    marginRight: 7,
  },
  backTextWhite: {
    color: "#FFF",
  },
  rowAlign: {
    flexDirection: "row",
  },
  rowFront: {
    paddingLeft: 20,
    backgroundColor: "#F1F1F0",
    borderBottomColor: "#DCDCDC",
    borderBottomWidth: 1,
    justifyContent: "center",
    height: 50,
  },
  notSelected:{
    backgroundColor: "#c9c7c7",
  },
  rowIcon: {
    marginRight: 12,
  },
  rowBack: {
    alignItems: "center",
    backgroundColor: "#F8F8FF",
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    paddingLeft: 15,
  },
  backRightBtn: {
    alignItems: "center",
    bottom: 0,
    justifyContent: "center",
    position: "absolute",
    top: 0,
    width: 75,
  },
  backRightBtnLeft: {
    backgroundColor: mainColor,
    right: 75,
  },
  backRightBtnRight: {
    backgroundColor: "red",
    right: 0,
  },
});
