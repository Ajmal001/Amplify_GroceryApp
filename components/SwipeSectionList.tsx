import React from "react";

import ListAccordion from "./ListAccordion";
import { mainColor, productCategory } from "../utils/helpers";
import { ScrollView, View } from "react-native";
import { FAB } from "react-native-paper";
import { SectionData } from "../src/types/ListElements"

interface Props {
  listData: SectionData[];
  deleteAction: () => void;
  navigateToEdit: () => void;
  onPressAction: () => void;
  toBuyView: boolean;
  fabAction: (text: string) => void;
  itemsInCart: boolean;
  groceryListID: string;
  expandedAll: boolean;
}


const SwipeSectionList: React.FC<Props> = ({
  listData,
  deleteAction,
  navigateToEdit,
  onPressAction,
  toBuyView,
  fabAction,
  groceryListID,
  itemsInCart,
  expandedAll
}) => {

  return (
    <View>
      {listData.map((section) => (
        <ListAccordion
          sectionTitle={section.title}
          sectionIcon={productCategory[section.key].picture}
          key={section.key}
          rowData={section.data}
          deleteAction={deleteAction}
          navigateToEdit={navigateToEdit}
          onPressAction={onPressAction}
          toBuyView={toBuyView}
          expandedAll={expandedAll}
        />
      ))}
         {toBuyView && listData.length > 0 && itemsInCart && (
        <FAB
          icon="clipboard-check-outline"
          label="I'm done!"
          style={{
            marginTop: 10,
            marginBottom: 10,
            backgroundColor: mainColor,
            alignSelf: "center",
          }}
          onPress={() => fabAction(groceryListID)}
        />
      )}
     
    </View>
  )
}

export default SwipeSectionList