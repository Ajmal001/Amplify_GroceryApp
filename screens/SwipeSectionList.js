import React, { useEffect, useState } from "react";
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableHighlight,
  View,
} from "react-native";
import { MaterialCommunityIcons } from "@expo/vector-icons";
import { SwipeListView } from "react-native-swipe-list-view";
import { blue, categoryPictures } from "../utils/helpers";
import { connect, useDispatch } from "react-redux";
import { handleDeleteProduct, handleLoadProducts } from "../src/redux/actions/product";

function SwipeSectionList(props) {
  const dispatch = useDispatch();
  const groceryListID = props.route.params.groceryList.id;
  useEffect(() => {
    dispatch(handleLoadProducts(groceryListID));
  }, []);
  const { listData } = props;

  const closeRow = (rowMap, rowKey) => {
    if (rowMap[rowKey]) {
      rowMap[rowKey].closeRow();
    }
  };

  const deleteRow = (rowMap, rowKey) => {
    closeRow(rowMap, rowKey);
    dispatch(handleDeleteProduct(rowKey));
  };
  const goToEditProduct = (rowMap, product) => {
    closeRow(rowMap, product.key);
    return props.navigation.push("AddProduct", { product });
  }

  const onRowDidOpen = (rowKey) => {
    console.log("This row opened", rowKey);
  };

  const renderItem = (data) => (
    <TouchableHighlight
      onPress={() => console.log("You touched me")}
      style={styles.rowFront}
      underlayColor={"#AAA"}
    >
      <View style={styles.rowAlign}>
        <MaterialCommunityIcons 
        name={data.item.checked ? 
          "checkbox-marked-circle"
          : "checkbox-blank-circle-outline"} size={20} color={blue} />
        <Text> {data.item.name} </Text>
      </View>
    </TouchableHighlight>
  );

  const renderHiddenItem = (data, rowMap) => (
    <View style={styles.rowBack}>
      <Text>Left</Text>
      <TouchableOpacity
        style={[styles.backRightBtn, styles.backRightBtnLeft]}
        onPress={() => goToEditProduct(rowMap, data.item)}
      >
        <MaterialCommunityIcons name="pencil" size={20} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[styles.backRightBtn, styles.backRightBtnRight]}
        onPress={() => deleteRow(rowMap, data.item.key)}
      >
        <MaterialCommunityIcons name="delete-outline" size={20} color="white" />
      </TouchableOpacity>
    </View>
  );

  const renderSectionHeader = ({ section }) => (
    <Text style={styles.sectionText}>
      <MaterialCommunityIcons
        name={categoryPictures[section.title]}
        size={20}
      />
      {section.title}
    </Text>
  );

  return (
    <View style={styles.container}>
      <SwipeListView
        useSectionList
        sections={listData}
        renderItem={renderItem}
        renderHiddenItem={renderHiddenItem}
        renderSectionHeader={renderSectionHeader}
        leftOpenValue={75}
        rightOpenValue={-150}
        previewRowKey={"0"}
        previewOpenValue={-40}
        previewOpenDelay={3000}
        onRowDidOpen={onRowDidOpen}
      />
    </View>
  );
}

function mapStateToProps(state) {
  const { products } = state;
  const currCategories = new Set();
  products.forEach((product) => currCategories.add(product.category));
  let currListCategories = Array.from(currCategories).map((cat) => ({
    title: cat,
    data: products
      .filter((product) => product.category === cat)
      .map((product) => ({ ...product, key: product.id })),
  }));

  return { listData: currListCategories };
}
export default connect(mapStateToProps)(SwipeSectionList);

const styles = StyleSheet.create({
  container: {
    backgroundColor: "white",
    flex: 1,
  },
  sectionText: {
    color: blue,
    fontWeight: "bold",
    marginLeft: 10,
  },
  backTextWhite: {
    color: "#FFF",
  },
  rowAlign:{
    flexDirection: 'row'
  },
  rowFront: {
    alignItems: "flex-start",
    paddingLeft: 20,
    backgroundColor: "#F8F8FF",
    borderBottomColor: "#DCDCDC",
    borderBottomWidth: 1,
    justifyContent: "center",
    height: 50,
  },
  rowBack: {
    alignItems: "center",
    backgroundColor: "#F8F8FF",
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    paddingLeft: 15,
  },
  backRightBtn: {
    alignItems: "center",
    bottom: 0,
    justifyContent: "center",
    position: "absolute",
    top: 0,
    width: 75,
  },
  backRightBtnLeft: {
    backgroundColor: "blue",
    right: 75,
  },
  backRightBtnRight: {
    backgroundColor: "red",
    right: 0,
  },
});
