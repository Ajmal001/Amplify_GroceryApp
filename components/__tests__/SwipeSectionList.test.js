import React from 'react';
import renderer from 'react-test-renderer';
import SwipeSectionList from '../SwipeSectionList';

jest.mock("@expo/vector-icons", () => {
    return {
      MaterialCommunityIcons: () => {
        return <></>;
      },
    };
  });


test('renders correctly', () => {
  const tree = renderer.create(<SwipeSectionList
    listData={[]}
    deleteProduct={x=>x}
    navigateToEditProduct={x=>x}
    toggleProduct={x=>x}
    toBuyView={true}/>).toJSON();
  expect(tree).toMatchSnapshot();
});