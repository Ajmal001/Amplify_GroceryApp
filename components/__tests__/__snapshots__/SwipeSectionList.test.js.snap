// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  closeOnRowBeginSwipe={false}
  closeOnRowOpen={true}
  closeOnRowPress={true}
  closeOnScroll={true}
  data={Array []}
  directionalDistanceChangeThreshold={2}
  disableHiddenLayoutCalculation={false}
  disableLeftSwipe={false}
  disableRightSwipe={false}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  leftOpenValue={75}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRowDidOpen={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  previewFirstRow={false}
  previewOpenDelay={3000}
  previewOpenValue={-40}
  previewRepeat={false}
  previewRepeatDelay={1000}
  previewRowKey="0"
  recalculateHiddenLayout={false}
  renderHiddenItem={[Function]}
  renderItem={[Function]}
  rightOpenValue={-150}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  swipeToClosePercent={50}
  swipeToOpenPercent={50}
  swipeToOpenVelocityContribution={0}
  updateCellsBatchingPeriod={50}
  useAnimatedList={false}
  useNativeDriver={true}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;
